crates/bevy_tasks/src/task_pool.rs

```diff
@@ -1,62 +1,35 @@
 use std::{
     future::Future,
     mem,
-    pin::Pin,
     sync::Arc,
     thread::{self, JoinHandle},
 };
 
-use futures_lite::{future, pin};
-
-use crate::Task;
-
-/// Used to create a [`TaskPool`]
-#[derive(Debug, Default, Clone)]
-#[must_use]
-pub struct TaskPoolBuilder {
-    /// If set, we'll set up the thread pool to use at most n threads. Otherwise use
-    /// the logical core count of the system
-    num_threads: Option<usize>,
-    /// If set, we'll use the given stack size rather than the system default
-    stack_size: Option<usize>,
-    /// Allows customizing the name of the threads - helpful for debugging. If set, threads will
-    /// be named <thread_name> (<thread_index>), i.e. "MyThreadPool (2)"
-    thread_name: Option<String>,
-}
-
-impl TaskPoolBuilder {
-    /// Creates a new [`TaskPoolBuilder`] instance
-    pub fn new() -> Self {
-        Self::default()
-    }
+pub use crate::task_pool_builder::*;
+use futures_lite::{future, FutureExt};
 
-    /// Override the number of threads created for the pool. If unset, we default to the number
-    /// of logical cores of the system
-    pub fn num_threads(mut self, num_threads: usize) -> Self {
-        self.num_threads = Some(num_threads);
-        self
-    }
+use crate::{Task, TaskGroup};
 
-    /// Override the stack size of the threads created for the pool
-    pub fn stack_size(mut self, stack_size: usize) -> Self {
-        self.stack_size = Some(stack_size);
-        self
-    }
+#[derive(Debug, Default)]
+struct GroupInfo {
+    executor: async_executor::Executor<'static>,
+    threads: usize,
+}
 
-    /// Override the name of the threads created for the pool. If set, threads will
-    /// be named `<thread_name> (<thread_index>)`, i.e. `MyThreadPool (2)`
-    pub fn thread_name(mut self, thread_name: String) -> Self {
-        self.thread_name = Some(thread_name);
-        self
-    }
+#[derive(Debug, Default)]
+struct Groups {
```

Comment:
Given that these are "context specific" groups and we might want more in the future (ex: pipelined rendering), I'm not seeing much justifying hard-coded Groups / TaskGroup enum types. It seems like we could make TaskPool generic on some TaskGroup type that maps to integer indices. Then we could use a `Vec<GroupInfo>` here and probably cut out a number of other hard-coded cases.

We could type alias the bevy task group to hide the generic. `type TaskPool = bevy_tasks::TaskPool<TaskGroup>`.

That being said, not necessarily worth blocking on.

Reply:
I don't think the current strategy will work if we genericize this over N possible executors. `future::or` is monomorphized, so we'd need to either:

 - Rip out the internals of async_executor and implement the priorization ourselves.
 - Eat the cost of using a `Pin<Box<dyn Future<Output = ()>>` to build a generic prioritization chain. This will probably exacerbate the scheduling overhead issues.

Reply:
Welp, ended up doing the first one here. Seems to work much better than the `future::or` solution as this allows for internal work-stealing.